### PR TITLE
[BEAM-77] Create runners/core module for artifact org.apache.beam:runners-core

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This is strictly creating the module and moving one easy class to it.
Many of the utilities in org.apache.beam.util and subpackages should
move as developments allow.

### CODE REVIEW

**Comment by kennknowles** [comment]
R: @davorbonaci @jbonofre 

I haven't worked too hard at the `pom.xml`. I also haven't gone through the loop of letting the dependency analysis clean up the deps, but this is the basic idea. Please comment!

File: runners/core-java/pom.xml [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+  <parent>
+    <groupId>org.apache.beam</groupId>
+    <artifactId>runners-parent</artifactId>
+    <version>0.1.0-incubating-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
+  </parent>
+
+  <artifactId>runners-core</artifactId>
+  <name>Apache Beam :: Runners :: Core</name>
+  <description>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: inline: <desc>Xyz</desc>

File: runners/core-java/pom.xml [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        <configuration>
+          <!-- Set testSourceDirectory in order to exclude generated-test-sources -->
+          <testSourceDirectory>${project.basedir}/src/test/</testSourceDirectory>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
+        <executions>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: rm

File: runners/core-java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+          </execution>
+        </executions>
+      </plugin>
+
+      <!-- Source plugin for generating source and test-source JARs. -->
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-source-plugin</artifactId>
+        <version>2.4</version>
+        <executions>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: rm

File: runners/core-java/pom.xml [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-javadoc-plugin</artifactId>
+        <configuration>
+          <windowtitle>Beam Runners Core utilities ${project.version} API</windowtitle>
+          <doctitle>Beam Runners Core utilities for Java, version ${project.version}</doctitle>
+          <overview>../javadoc/overview.html</overview>
+
+          <subpackages>org.apache.beam.runners.core</subpackages>
+          <additionalparam>-exclude org.apache.beam.sdk.runners.worker:org.apache.beam.sdk.runners.dataflow:org.apache.beam.sdk.util:org.apache.beam.sdk.runners.inprocess ${dataflow.javadoc_opts}</additionalparam>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: Will we need to exclude these? I'd remove.

File: runners/core-java/pom.xml [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+
+      <!-- This plugin tells Maven about an additional test-source directory to
+           build, which contains Avro-generated source files. This is not
+           strictly needed for the regular Maven build, but helps certain IDEs
+           automatically find and compile generated code. -->
+      <plugin>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: likely won't be needed here

File: runners/core-java/pom.xml [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>java-sdk-all</artifactId>
+      <version>${project.version}</version>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: rm all versions

File: runners/direct-java/pom.xml [hunk 7]
```diff
@@ -243,6 +243,27 @@
     <dependency>
       <groupId>org.apache.beam</groupId>
       <artifactId>java-sdk-all</artifactId>
+      <version>${project.version}</version>
```
Comment (davorbonaci) [inline runners/direct-java/pom.xml]: rm all versions

File: runners/core-java/pom.xml (hunk 1 above)
Comment (kennknowles) [inline runners/core-java/pom.xml]: Does that matter? The formatting gets bad.

Comment (kennknowles) [inline runners/core-java/pom.xml]: Nevermind; short enough for one line.

**Comment by kennknowles** [comment]
Please take another look when you get a chance. The two commits are independent, not to be squashed.

File: runners/core-java/pom.xml [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-all</artifactId>
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>provided</scope>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: scope test?

File: runners/core-java/pom.xml [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+      <artifactId>java-sdk-all</artifactId>
+    </dependency>
+
+    <!-- build dependencies -->
+
+    <!-- test dependencies -->
+    <dependency>
+      <groupId>org.hamcrest</groupId>
+      <artifactId>hamcrest-all</artifactId>
+      <scope>provided</scope>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: scope test?

File: runners/core-java/pom.xml [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+      <!-- Coverage analysis for unit tests. -->
+      <plugin>
+        <groupId>org.jacoco</groupId>
+        <artifactId>jacoco-maven-plugin</artifactId>
+      </plugin>
+
+      <!-- Avro plugin for automatic code generation -->
+      <plugin>
+        <groupId>org.apache.avro</groupId>
+        <artifactId>avro-maven-plugin</artifactId>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: I don't think this will be needed. Cut this plugin?

File: runners/core-java/pom.xml [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+          </execution>
+        </executions>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-checkstyle-plugin</artifactId>
+        <configuration>
+          <!-- Set testSourceDirectory in order to exclude generated-test-sources -->
+          <testSourceDirectory>${project.basedir}/src/test/</testSourceDirectory>
```
Comment (davorbonaci) [inline runners/core-java/pom.xml]: I don't think we'll have any generated test sources here. (That is only for AvroIO/AvroCoder-related tests, I think.) So, this might not be needed.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/DoFnRunner.java [hunk 12]
```diff
@@ -37,8 +37,8 @@
   public void processElement(WindowedValue<InputT> elem);
 
   /**
-   * Calls {@link DoFn#finishBundle} and performs additional tasks, such as
-   * flushing in-memory states.
+   * Calls {@link DoFn#finishBundle} and performs additional tasks, such as flushing in-memory
```
Comment (davorbonaci) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/DoFnRunner.java]: Revert this file? Looks unrelated.

File: runners/flink/runner/pom.xml [hunk 13]
```diff
@@ -101,6 +101,17 @@
 
     <dependency>
       <groupId>org.apache.beam</groupId>
+      <artifactId>runners-core</artifactId>
```
Comment (davorbonaci) [inline runners/flink/runner/pom.xml]: Add Spark dependency too?

Comment (davorbonaci) [inline runners/flink/runner/pom.xml]: As well as Dataflow runner?
(unless it breaks dep check)

**Comment by davorbonaci** [comment]
LGTM, modulo resolution of pending comments. (I probably don't need to take another look, unless there's something contentious.)

File: runners/flink/runner/pom.xml (hunk 13 above)
Comment (aljoscha) [inline runners/flink/runner/pom.xml]: Why add them to the Flink runner?

Comment (kennknowles) [inline runners/flink/runner/pom.xml]: Short answer: Flink runner uses `GroupAlsoByWindowViaWindowSetDoFn`.

Long answer: The intention for `runners/core` is to contain all the code from the SDK that is not relevant for users, but just for executing a pipeline. In particular, all the GABW implementations, the `GroupByKeyViaGroupByKeyOnly` helper, most `DoFnRunner` implementations, `SideInputReader` and implementations, and `ReduceFnRunner`. It should be a good majority of the SDK's `util` package.

I deliberately chose to move the class I did just to establish a couple dependencies so this new module wasn't entirely vacuous. Spark and Flink batch do not yet use this, but I imagine they will later.

Comment (aljoscha) [inline runners/flink/runner/pom.xml]: Ah, I was talking about @davorbonaci's comment about adding a Spark dep and a data flow dep.

The intention of the other stuff was pretty clear. 😃

File: runners/core-java/pom.xml (hunk 8 above)
Comment (kennknowles) [inline runners/core-java/pom.xml]: Done.

File: runners/core-java/pom.xml (hunk 9 above)
Comment (kennknowles) [inline runners/core-java/pom.xml]: Done. Note that things like `CoderProperties` made it `provided` for the SDK. For `runners-core` it can be `test` until we have test utilities we offer.

File: runners/core-java/pom.xml (hunk 10 above)
Comment (kennknowles) [inline runners/core-java/pom.xml]: Done.

File: runners/core-java/pom.xml (hunk 11 above)
Comment (kennknowles) [inline runners/core-java/pom.xml]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/util/DoFnRunner.java (hunk 12 above)
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/util/DoFnRunner.java]: Done.
